src/Payments.sol

```diff
@@ -0,0 +1,612 @@
+                "failed to settle: arbiter modified amount exceeds maximum for settled duration"
+            );
+
+            // Adjust the amount and the final epoch according to the arbiter's decision.
+            settledUntil = arbResult.settleUpto;
+            settledAmount = arbResult.modifiedAmount;
+            note = arbResult.note;
+        }
+
+        // Revert if no progress was made (i.e., the final epoch did not advance).
+        require(settledUntil > epochStart, "failed to settle: no progress made in settlement");
+
+        return (settledAmount, settledUntil, note);
+    }
+
+    function _settleWithRateChanges(
+        uint256 railId,
+        Rail storage rail,
+        RateChangeQueue.Queue storage rateQueue,
+        uint256 initialRate,
+        uint256 startEpoch,
+        uint256 targetEpoch
+    ) internal returns (uint256 totalSettled, uint256 finalEpoch, string memory note) {
+        totalSettled = 0;
+        uint256 currentEpoch = startEpoch;
+        uint256 activeRate = initialRate;
+        note = "";
+
+        while (currentEpoch <= targetEpoch) {
+            uint256 nextBoundary = targetEpoch;
+
+            // If there's an upcoming rate change, check if it applies within our current range.
+            if (!rateQueue.isEmpty()) {
+                RateChangeQueue.RateChange memory upcomingChange = rateQueue.peek();
+                bool isWithinRange = (
+                    upcomingChange.untilEpoch >= currentEpoch &&
+                    upcomingChange.untilEpoch <= targetEpoch
+                );
+                require(isWithinRange, "rate queue is in an invalid state");
+                nextBoundary = upcomingChange.untilEpoch;
+                activeRate = upcomingChange.rate;
+            }
+
+            // Settle the segment from the current epoch up to the next boundary (or early if arbitration says so).
+            (uint256 segmentAmount, uint256 segmentEndEpoch, string memory arbNote) = _settleSegment(
+                railId,
+                rail,
+                currentEpoch,
+                nextBoundary,
+                activeRate
+            );
+
+            // Update the total settled. 
+            totalSettled += segmentAmount;
+
+            // If the arbitration shortened the segment, stop and return immediately.
+            // but keep the rate change in the queue as we've not fully settled the segment.
+            if (segmentEndEpoch < nextBoundary) {
+                return (totalSettled, segmentEndEpoch, arbNote);
+            }
+
+            // Otherwise, we have fully settled up to nextBoundary.
+            currentEpoch = segmentEndEpoch;
+            note = arbNote;
+
+            // Dequeue the rate change we just passed (if it was indeed used this round).
+            if (!rateQueue.isEmpty()) {
+                // Peek again to confirm it's the one we used.
+                RateChangeQueue.RateChange memory frontChange = rateQueue.peek();
```

Comment:
How would this not be the case?

Reply:
if the queue is non-empty AND next element has expired -> until epoch is in the past or < targetEpoch -> pop.

Reply:
actually -> we can JUST dequeue if we reach here.

Reply:
Done.